### ISSUE TITLE
Connectivity Issues

### ISSUE THREAD

**Issue by oneneo11** [spagettikod/gotracer#1]
Hi

When I run the example code, I get the following:

C:\Users\sidg\Documents\gotracer-master>go run abc.go
ArrayVoltage: 0.00
ArrayCurrent: 0.00
ArrayPower: 0.00
BatteryVoltage: 0.00
BatteryCurrent: 0.00
BatterySOC: 0%
BatteryTemp: 0.00
BatteryMaxVoltage: 0.00
BatteryMinVoltage: 0.00
DeviceTemp: 0.00
LoadVoltage: 0.00
LoadCurrent: 0.00
LoadPower: 0.00
Load: false
EnergyConsumedDaily: 0.00
EnergyConsumedMonthly: 0.00
EnergyConsumedAnnual:0.00
EnergyConsumedTotal:0.00
EnergyGeneratedDaily: 0.00
EnergyGeneratedMonthly: 0.00
EnergyGeneratedAnnual: 0.00
EnergyGeneratedTotal: 0.00

How do you think I can troubleshoot this?

Thanks

**Comment by spagettikod** [comment]
Did you get the communication to work with the software from EPSolar on the port?

**Comment by oneneo11** [comment]
Yes I was able to connect to the EPsolar monitor software.

**Comment by spagettikod** [comment]
Hmm, the only thing I can come up with is if EPSolar changed the protocol on newer Tracers, mine is about 18 months old. I have access to a Windows computer in about 3 weeks, I can try it out and see if it works on my Tracer.

**Comment by oneneo11** [comment]
I finally got it to work but I had to use https://github.com/jacobsa/go-serial to access my COM ports. Everything else works fine.

BTW

```
queryStateCommand []command = []command{{data: []byte{0x01, 0x04, 0x32, 0x00, 0x00, 0x03, 0xbe, 0xb3}, respLen: 11, offset: 0},
    {data: []byte{0x01, 0x02, 0x20, 0x00, 0x00, 0x01, 0xb2, 0x0a}, respLen: 6, offset: 11},
    {data: []byte{0x01, 0x43, 0x31, 0x00, 0x00, 0x1b, 0x0a, 0xf2}, respLen: 51, offset: 17},
    {data: []byte{0x01, 0x04, 0x33, 0x1a, 0x00, 0x03, 0x9e, 0x88}, respLen: 11, offset: 68},
    {data: []byte{0x01, 0x04, 0x33, 0x02, 0x00, 0x12, 0xde, 0x83}, respLen: 41, offset: 79}}
```

Should the respLen = 7 for the second line above and respLen = 59 on the third line.

**Comment by spagettikod** [comment]
Okey. I changed to the other one to get it to work in Linux. :)

Do you mean it doesn't work if you don't change to 7 and 59? I think I'm getting correct readouts with the current values.

**Comment by oneneo11** [comment]
Cool, maybe you could put in a disclaimer to use the other go-serial for Windows. 

Sorry the respLen = 6 is correct. Because the there is only 1 byte of information replied back.

But the respLen of 59 just adds more information from the register. You are telling it to respond from 3100 to 311D (0x001D stands for the count of the address). Each address I think is supplying with 2 bytes of data. 

receive: 
**first response**: 01 04 06 00 00 00 01 00 01 F0 93 
**second response**: 01 02 01 00 A1 88 
**third response**: 01 43 36 04 0F FF FF 00 07 00 00 00 00 00 00 04 C7 00 00 00 00 00 00 04 C7 00 00 00 00 00 00 04 C7 00 00 00 00 00 00 09 C4 09 DB 09 DB 09 C4 00 1A 1F FC 00 00 00 00 00 00 00 00 
**fourth response**: 01 04 06 04 C7 00 00 00 00 D4 C6 
**fifth response**: 01 04 24 04 CA 04 C5 00 00 00 00 00 00 00 00 00 00 00 00 00 16 00 00 00 00 00 00 00 00 00 00 00 00 00 00 01 29 00 00 38 D1

If you look at the third response, the actual raw data will say something as 01 43 36....  
-> 0x01 = device id
-> 0x43 = register
-> 0x36 = count of the following response information (DEC of 36 = 54)
-> 0xD4C6 = crc

The total response length is 54+5 = 59. The 5 count corresponds to the first three responses: device id, register, count of information in the response and the two crcs at the end of the response.

**Comment by spagettikod** [comment]
Yes I will need to mention that. The project I'm using is supposed to work on Windows too but I might be doing something wrong. Will try to replicate when I get the time. 

Don't know why I skip the last parts of the response. I remember having a hard time finding information about the protocol and deciphered a lot by listening to the communication between their software and the Tracer. Did you change it to 59 and tried it?

**Comment by oneneo11** [comment]
Maybe I did something and the old go-serial package did not like it. 

Yup. I tried it and works fine. 

I am using your file to log from the solar controller into a txt file.

Sometimes I get bad data like this 172.06,10.39,655.35,0,0,35,0,12.89,12.84,0,0,0,0,0,0,0,0.22,0,0.79,0.79,3.76

It looks like there is a bug in the software of the controller where it sends the variables before they are properly calculated.  It does not happen consistently. About 10% of the time. I think a crc check could be included to ignore these strings of data. But I dont know how the crc is calculated by this controller.

I dont think they follow the standard MODBUS crcIBM method but I could be wrong and lost.
